File: src/Csharp/packages/geometry/src/OBJLoader.cs
using System;
using System.Text.RegularExpressions;

namespace DHARTAPI.Geometry
{


	/*! 
	  \brief Methods for seperating meshes when loading from OBJ
	
	*/
	public enum GROUP_METHOD{
		ONLY_FILE = 0, ///< Don't group submeshes. Return the entire file as a single MeshInfo.
		BY_GROUP = 1, ///< Create a seperate meshinfo for every obj group in the file.
		BY_MATERIAL = 2 ///< Create a seperate meshinfo for every material in the file. Requires the .mtl to be located next to the OBJ upon loading. If a .mtl could not be found, ONLY_FILE is used as a fallback.
	}

    /*! 
        \brief Load mesh geometry from OBJ files on disk.

        \remarks
        Internally is doing the same work as calling MeshInfo's constructor, so loaders
        for other filetypes can be created.
    */
    public static class OBJLoader
    {
        /*!
            \brief Load an obj from the OBJ file at the given filepath.

			\param path Path to a valid OBJ file.
			\param xrot Degrees to rotate the mesh on the x axis.
			\param yrot Degrees to rotate the mesh on the y axis.
			\param zrot Degrees to rotate the mesh on the z axis.
			
			\returns An instance of MeshInfo containing the mesh loaded from the OBJ at path.

            \throws System.IO.FileNotFoundException No file was found at path.
			\throws DHARTAPI.Exceptions.InvalidMeshException
			The file at the path was not a valid OBJ.
			
            \remarks
            Use the other overload if you want to use a CommonRotation on the mesh. 

			\code
				// Loading a mesh from plane.obj with no rotation.
				MeshInfo MI = OBJLoader.LoadOBJ("plane.obj");

			\endcode
        */
        public static MeshInfo LoadOBJ(string path, float xrot = 0, float yrot = 0, float zrot = 0)
        {
            var mesh_ptr = NativeMethods.C_LoadOBJ(path, xrot, yrot, zrot, GROUP_METHOD.ONLY_FILE);
            return new MeshInfo(mesh_ptr[0]);
		}

		/*!
			\brief Load an obj from the OBJ file at the given filepath.

			\param path Path to a valid OBJ file.
			\param gm Method of grouping the OBJ into submeshes
			\param xrot Degrees to rotate the mesh on the x axis.
			\param yrot Degrees to rotate the mesh on the y axis.
			\param zrot Degrees to rotate the mesh on the z axis.

			\returns All submeshes in the obj file at `path` grouped by `gm`

			\throws System.IO.FileNotFoundException No file was found at  path.
			\throws DHARTAPI.Exceptions.InvalidMeshException
			The file at the path was not a valid OBJ.
			
			\see GROUP_METHOD to see the different ways of grouping the different geometry in a .obj file

			\par Example
			\snippet geometry\test_geometry.cs EX_LoadSubmeshes

			```
			Name: arcs_floor, ID: 0, Verts: 150, Triangles: 192
			Name: arcs_03, ID: 1, Verts: 750, Triangles: 1344
			Name: object32, ID: 2, Verts: 3190, Triangles: 4784
			Name: object31, ID: 3, Verts: 3454, Triangles: 5264
			Name: pillar_cor, ID: 4, Verts: 3478, Triangles: 5312
			```
			\code
			\endcode
		*/
		public static MeshInfo[] LoadOBJSubmeshes(string path, GROUP_METHOD gm, float xrot = 0, float yrot = 0, float zrot = 0)
        {
			// Load from native memory
            IntPtr[] mesh_ptrs = NativeMethods.C_LoadOBJ(path, xrot, yrot, zrot, gm);

			// Iterate through each submesh and create new MeshInfo objects for each.
			int num_meshes = mesh_ptrs.Length;
			MeshInfo[] return_meshes = new MeshInfo[num_meshes];
			for (int i = 0; i < num_meshes; i++)
				return_meshes[i] = new MeshInfo(mesh_ptrs[i]);

			return return_meshes;
		}

		/*!
            \brief Load an obj from the OBJ file at the given filepath.

			\param path Path to a valid OBJ file.
			\param rotation Degrees to rotate the mesh in the x,y, and z direction after loading. 

			\returns An instance of MeshInfo containing the mesh loaded from the OBJ at path.

            \throws System.IO.FileNotFoundException No file was found at  path.
			\throws DHARTAPI.Exceptions.InvalidMeshException
			The file at the path did not represent a valid OBJ.
			
            \remarks
			The rotation parameter can be used with the rotations in CommonRotations to easily
			perform common rotations on meshes. This is useful if you're loading a mesh that is
			Y-up and you need to convert it to Z-up for use in the GraohGenerator. 
		
			\endinternal

			\code
				// Loading a mesh from plane.obj with no rotation.
				MeshInfo MI = OBJLoader.LoadOBJ("plane.obj");
			\endcode
		*/
		public static MeshInfo LoadOBJ(string path, Vector3D rotation) => LoadOBJ(path, rotation.x, rotation.y, rotation.z);
    }
}

File: src/Python/dhart/geometry/obj_loader.py
from typing import Tuple, Union, List
from enum import IntEnum

from . import meshinfo_native_functions
from . import MeshInfo

__all__ = ['OBJGroupType','LoadOBJ']

class OBJGroupType(IntEnum):
    """ Different methods to seperate several meshes from a single OBJ """
    ONLY_FILE = 0  # Assign the same ID to all geometry in this file
    BY_GROUP = 1  # Assign different OBJ groups, different OBJs
    BY_MATERIAL = 2  # Seperate geometry based on material


def LoadOBJ(
    path: str,
    group_type: OBJGroupType = 0,
    rotation: Tuple[float, float, float] = (0, 0, 0),
) -> Union[MeshInfo, List[MeshInfo]]:
    """ Load an obj file from the given path. 
    
    Args:
        path: The path to a valid OBJ file.
        group_type: (NOT IMPLEMENTED) How IDs will be assigned to different
            parts of the mesh 
        rotation: Rotation in degrees to be performed on the OBJ after it is loaded. 
            Useful for converting models from Y-Up to Z-Up.
    
    Returns:
        MeshInfo: A new meshinfo object containing a the vertices and triangles
            for the obj in path. If group_type is not ONLY_FILE, a list of meshes
            may be returned if submeshes are found. 

    Raises:
        dhart.Exceptions.InvalidOBJException: The OBJ at path
            either did not exist or could not be loaded
        dhart.Exceptions.FileNotFoundException: No file exists
            at the given path.

    Example:
        Load plane.obj from a folder titled "Example Models" but fail because that file doesn't exist
        
        >>> import dhart
        >>> from dhart.geometry import LoadOBJ
        >>> obj = dhart.get_sample_model("not_real.obj")
        >>> MI = LoadOBJ(obj)
        Traceback (most recent call last):
        ...
        dhart.Exceptions.FileNotFoundException

        >>> import dhart
        >>> from dhart.geometry import LoadOBJ
        >>> obj = dhart.get_sample_model("plane.obj")
        >>> MI = LoadOBJ(obj)
        >>> MI 
        (EntireFile, 0)


    """

    mesh_ptr = meshinfo_native_functions.CreateOBJ(path, group_type, rotation)

    # Check return type. If multiple meshes were reutrned, then make a list
    if (not isinstance(mesh_ptr, list)):
        return MeshInfo(mesh_ptr)
    else:
        return [MeshInfo(ptr) for ptr in mesh_ptr]